consensus/fork.go

```diff
@@ -177,13 +177,27 @@ func (s *State) forkBlockchain(newNode *BlockNode) (err error) {
 	var appliedNodes []*BlockNode
 	for i := len(backtrackNodes) - 1; i >= 0; i-- {
 		appliedNodes = append(appliedNodes, backtrackNodes[i])
+
+		// If the diffs for this node have already been generated, apply them
+		// directly instead of generating them.
+		if backtrackNodes[i].DiffsGenerated {
+			for _, outputDiff := range backtrackNodes[i].OutputDiffs {
+				s.commitOutputDiff(outputDiff, true)
+			}
+			for _, contractDiff := range backtrackNodes[i].ContractDiffs {
+				s.commitContractDiff(contractDiff, true)
+			}
+			continue
+		}
+
+		// If the diffs have not been generated, call generateAndApply.
 		err = s.generateAndApplyDiff(backtrackNodes[i])
 		if err != nil {
 			// Invalidate and delete all of the nodes after the bad block.
 			s.invalidateNode(backtrackNodes[i])
 
 			// Rewind the validated blocks
-			for j := 0; j < i; j++ {
+			for j := 0; j < len(appliedNodes)-1; j++ {
```

Comment:
I feel like there's a better way to write this... `for range appliedNodes {` would *almost* work, but you want one iteration fewer than that. Maybe `for range appliedNodes[1:] {`.

Alternatively, use a function called `s.invertBlock(Block)`, or at least rename this one to `s.invertLastBlock()`.